packages/SwingSet/docs/garbage-collection.md

```diff
@@ -519,3 +474,121 @@ The delivery of each GC action is processed as follows:
+
+Alternatively, the exporter might retire the object itself (typically when the exporter does not have any internal references to the object, so an incoming `dropExport` causes the object to be deleted entirely). When this happens, a cascade of `retireImport` messages will flow *downstream* towards the importers.
+
+These two kinds of messages might cross in the middle, so we must accomodate a variety of exceptional conditions where they meet.
+
+## Comms GC Retirement (upstream)
+
+If Vat-3 is not using a given Presence as a WeakMap key, then when it drops the object, it will immediately retire it as well. The vat will perform `syscall.dropImport` and `syscall.retireImport` in the same crank. This results in two messages being sent to the local comms vat: `dispatch.dropExport` and `dispatch.retireExport`. If instead the object *was* in use as a WeakMap key, the `retire` will not happen unless/until that WeakMap is deleted. So it could occur much later, or not at all.
+
+The importer's retirement causes a sequence of upstream `retireExport` messages to be sent:
+
+![Comms Retire Sequence (upstream) diagram](./images/gc/comms-retire-sequence-upstream.png "Comms Retire Sequence (upstream)")
+
+## Comms GC Retirement (downstream)
+
+Once the exporting vat has received the drop, liveslots deletes the strong reference it held on the Remotable. If nothing else in the vat has an additional strong reference, the Remotable will be garbage-collected and its identity deleted, triggering a downstream retirement message. If the object is actually a "virtual object", then a refcount is adjusted, and the virtual object might be released if it reaches zero, triggering similar downstream retirement.
+
+The upstream comms vat will send `remote.retireImport` messages to its downstream importing systems. This will eventually cause the final importing comms vat to do a `syscall.retireExport` (since that comms vat was exporting the object into the kernel), which eventually makes that kernel do `dispatch.retireImport` into the vat. This will allow the downstream vat to delete the WeakMap/WeakSet entry, releasing any data used as a value.
+
+The sequence of downstream messages looks like this:
+
+![Comms Retire Sequence (downstream) diagram](./images/gc/comms-retire-sequence-downstream.png "Comms Retire Sequence (downstream)")
+
+## Informed vs Ignorant, Seqnum and Acknum
+
+Despite our best efforts, the speed of light is not infinite, and messages sent from one comms vat to another will not arrive immediately. Machine A might send a message to machine B about some object which B is trying to drop, and these two messages might "cross on the wire". We use sequence and acknowledgment numbers to keep track of who knew what, and when, so we can properly handle the race between these crossing messages.
```

Comment:
We should try to do something about the speed of light once we have the resources.